promql/engine.go

```diff
@@ -1530,11 +1543,15 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
-		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
-			newEv.startTimestamp += newEv.interval
+		if !e.AlignEvalTime {
+			// Start with the first timestamp after (ev.startTimestamp - offset - range)
+			// that is aligned with the step (multiple of 'newEv.interval').
+			newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
+			if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+				newEv.startTimestamp += newEv.interval
+			}
+		} else {
+			newEv.startTimestamp = ev.startTimestamp - offsetMillis - rangeMillis
```

Comment:
This PR still looks wrong, this code is far too simple to implement this correctly as it's only looking at the timestamp of the first step - not all steps. Thus any other steps for which it is correct are only by chance, indicating that there aren't enough tests for this either as they're not catching this. It's also not testing this nested, and you're possibly also leaking implementation details of how subquery start times are determined for that case (from a semantic standpoint, subqueries don't have start times, it's just that we only request some of it for efficiency).

Keep in mind that if the solution looks simple or efficient, then it is wrong. A correct solution has to allow for potentially a different alignment for every single step in the worst case, and thus would require multiple range evals to cover each of these alignments.

I would suggest that if you want to make this work in realistic way, have the user pass in a static literal alignment. That way it's the same for all steps, and you can avoid the above fun.